api: first pass at configuration object

squeed opened the pull request.

rajatchopra commented on pkg/apis/networkoperator/v1/types.go:
```diff
@@ -4,26 +4,110 @@ import (
+	// or similar functionality. If unset, the plugin will attempt to select
+	// the correct value.
+	// +optional
+	DeployKubeProxy *bool
```
Or should there be another operator for kube-proxy?

squeed replied:
I don't think so, no. It's just one more manifest to deploy.

squeed commented:
@danwinship thanks a lot for the feedback, PTAL. I've made a few updates:

- clarified how the "NetworkDefition" struct is supposed to work w.r.t the optional Config fields. I organized it better too
- Added types + enums for all of the non-free strings
- Removed the `Primary` attribute, added `DefaultNetwork` and `AdditionalNetworks`

squeed commented:
I still need to figure out the KubeProxyConfig type

danwinship commented on pkg/apis/networkoperator/v1/types.go:
```diff
@@ -4,26 +4,148 @@ import (
+type SDNMode string
+
+const (
+	SDNModeMultitenant SDNMode = "multitenant"
```
normally in kubernetes, string-valued "enum" types have PascalCased values, not all-lowercase...

squeed replied:
Oh duh.

danwinship commented on pkg/apis/networkoperator/v1/types.go:
```diff
@@ -4,26 +4,148 @@ import (
+// Kind must be specified, along with exactly one "Config" that matches
+// the kind. Kinds that do not have a specific configuration parameter should
+// use OtherConfig
+type NetworkDefinition struct {
```
I think it's weird to use the same type for DefaultNetwork and AdditionalNetworks; the default network doesn't need a name and can't use RawCNIConfig, whereas the additional networks require a name, and can *only* be RawCNIConfig. So it seems like you could have something like:

    // DefaultNetworkDefinition represents the default network plugin's configuration. Exactly one
    // of the three fields must be set
    type DefaultNetworkDefinition struct {
            OpenshiftSDNConfig *OpenshiftSDNConfig
            OVNKubernetesConfig *OVNKubernetesConfig
            OtherConfig *map[string]string
    }

    type AdditionalNetworkDefinition struct {
            Name string
            RawCNIConfig string
    }

squeed replied:
That's a good point, but there's nothing saying that this always has to be the case. It's quite probable that we will have a supported network provider (maybe even with its own config type) that could be either primary or additional. Ideally they render the same (with, perhaps, different parameters).

We definitely could have someone using RawCNIConfig as a primary network configuration (though this would indeed have to be rendered to disk, not a CRD).

I think you're right, though, that two separate types could make it more user-friendly, even if they look very similar. Let me updated it and let's see how it looks.
